libp2p-identify/src/transport.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::{stream, Future, IntoFuture, Stream};
+use libp2p_peerstore::{PeerAccess, PeerId, Peerstore};
+use libp2p_swarm::{MuxedTransport, Transport};
+use multiaddr::{AddrComponent, Multiaddr};
+use protocol::{IdentifyInfo, IdentifyOutput, IdentifyProtocolConfig};
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
+use std::ops::Deref;
+use std::time::Duration;
+
+/// Implementation of `Transport`. See the crate root description.
+#[derive(Debug, Clone)]
+pub struct IdentifyTransport<T, P> {
+	transport: T,
+	peerstore: P,
+	addr_ttl: Duration,
+}
+
+impl<T, P> IdentifyTransport<T, P> {
+	/// Creates an `IdentifyTransport` that wraps around the given transport and peerstore.
+	#[inline]
+	pub fn new(transport: T, peerstore: P) -> IdentifyTransport<T, P> {
+		IdentifyTransport::with_ttl(transport, peerstore, Duration::from_secs(3600))
+	}
+
+	/// Same as `new`, but allows specifying a time-to-live for the addresses gathered from
+	/// remotes that connect to us.
+	///
+	/// The default value is one hour.
+	#[inline]
+	pub fn with_ttl(transport: T, peerstore: P, ttl: Duration) -> IdentifyTransport<T, P> {
+		IdentifyTransport {
+			transport: transport,
+			peerstore: peerstore,
+			addr_ttl: ttl,
+		}
+	}
+}
+
+impl<T, Pr, P> Transport for IdentifyTransport<T, Pr>
+where
+	T: Transport + Clone + 'static,          // TODO: 'static :(
```

Comment:
For nontrivial uses of generics can we give proper names to the type parameters like we would arguments? You wouldn't get away with naming nontrivial arguments `a` and `b`, why should type parameters get a free pass? My personal choice would be `impl<Trans, PStoreRef, PStore>`. Also, ideally `PStore` would be before `PStoreRef` and the `where` bounds reordered accordingly since that makes it more readable in my opinion.

Reply:
That's true, it's a bad habit I have now.